src/Cli/Microsoft.DotNet.Cli.Utils/MSBuildArgs.cs

```diff
@@ -0,0 +1,155 @@
+        var restoreProperties = parseResult.GetValue(restorePropertiesOption);
+        var otherMSBuildArgs = parseResult.UnmatchedTokens.ToArray();
+        return new MSBuildArgs(
+            properties: globalProperties,
+            restoreProperties: restoreProperties,
+            otherMSBuildArgs: otherMSBuildArgs);
+    }
+
+    public static MSBuildArgs FromProperties(FrozenDictionary<string, string>? properties)
+    {
+        return new MSBuildArgs(properties, null, []);
+    }
+
+    public static MSBuildArgs ForHelp = new(null, null, ["--help"]);
+
+    public MSBuildArgs CloneWithExplicitArgs(string[] newArgs)
+    {
+        return new MSBuildArgs(
+            properties: GlobalProperties,
+            restoreProperties: RestoreGlobalProperties,
+            otherMSBuildArgs: newArgs);
+    }
+
+    public MSBuildArgs CloneWithAdditionalRestoreProperties(FrozenDictionary<string, string>? additionalRestoreProperties)
+    {
+        if (additionalRestoreProperties is null || additionalRestoreProperties.Count == 0)
+        {
+            // If there are no additional restore properties, we can just return the current instance.
+            return new MSBuildArgs(GlobalProperties, RestoreGlobalProperties, OtherMSBuildArgs.ToArray());
+        }
+        if (RestoreGlobalProperties is null)
+        {
+            return new MSBuildArgs(GlobalProperties, additionalRestoreProperties, OtherMSBuildArgs.ToArray());
+        }
+
+        var newRestoreProperties = new Dictionary<string, string>(RestoreGlobalProperties, StringComparer.OrdinalIgnoreCase);
+        foreach (var kvp in additionalRestoreProperties)
+        {
+            newRestoreProperties[kvp.Key] = kvp.Value;
+        }
+        return new MSBuildArgs(GlobalProperties, newRestoreProperties.ToFrozenDictionary(newRestoreProperties.Comparer), OtherMSBuildArgs.ToArray());
+    }
+
+    public MSBuildArgs CloneWithAdditionalProperties(FrozenDictionary<string, string>? additionalProperties)
+    {
+        if (additionalProperties is null || additionalProperties.Count == 0)
+        {
+            // If there are no additional properties, we can just return the current instance.
+            return new MSBuildArgs(GlobalProperties, RestoreGlobalProperties, OtherMSBuildArgs.ToArray());
+        }
+        if (GlobalProperties is null)
+        {
+            return new MSBuildArgs(additionalProperties, RestoreGlobalProperties, OtherMSBuildArgs.ToArray());
+        }
+
+        var newProperties = new Dictionary<string, string>(GlobalProperties, StringComparer.OrdinalIgnoreCase);
+        foreach (var kvp in additionalProperties)
+        {
+            newProperties[kvp.Key] = kvp.Value;
+        }
+        return new MSBuildArgs(newProperties.ToFrozenDictionary(newProperties.Comparer), RestoreGlobalProperties, OtherMSBuildArgs.ToArray());
+    }
+
+    public void ApplyPropertiesToRestore()
```

Comment:
Why can't we do this always at initialization time?

Reply:
Because we don't know how the args are going to be used. We only need to set RestoreProperties if the command has the `-restore` flag and doesn't have one of the properties that cause us to trigger a separate restore. That logic doesn't live in this type (yet?).

Because the MSBuildForwardingAppWithoutLogging unconditionally emits this entire MSBuildArgs structure as args, we have to make sure that we don't call this unless we're sure we're in the codepath that needs restore properties.